### ISSUE TITLE
natvis does not seems to work correctly with gdb

### ISSUE THREAD

**Issue by UR3IQO** [microsoft/vscode-cpptools#476]
Hi, All!

I am using cpptools 0.10.0 and vscode 1.9.0. I use vscode for embedded development and debug with  gdb (arm-none-eabi-gdb). Debugging works finely! I tried to add some visualization using .natvis file, but it seems visualization works only partially. For example I was able to unroll the array using _ArrayItems_, but I am unable to just display string using _DisplayString_ or create synthetic item using _Synthetic_. 

For example this .natvis file piece:
```
  <Type Name="Bitmap">
      <DisplayString>{width} x {height}</DisplayString>
      <Expand>
         <Item Name="data size">width*height</Item>
         <ArrayItems>
            <Size>width*height</Size>
            <ValuePointer>data</ValuePointer>
         </ArrayItems>
      </Expand>
  </Type>

```
correctly shows _Item_ as well as _ArrayItems_ , but ignores _DisplayString_.

**Comment by pieandcakes** [comment]
@sov1178 Do you have a code snippet that I can apply this to for investigation?

**Comment by pieandcakes** [comment]
@sov1178 Also, can you make sure you have the property `"showDisplayString": true` set in your launch.json?

**Comment by UR3IQO** [comment]
@pieandcakes  
> can you make sure you have the property "showDisplayString": true set in your launch.json?

Thanks a lot! Adding _showDisplayString_ to launch.json solved described _DisplayString_ problem. I think it should be added tho the launch.json documentation ( https://github.com/Microsoft/vscode-cpptools/blob/master/launch.md ).

**Comment by UR3IQO** [comment]
But adding _Condition_ attribute to _DisplayString_ element do not produce expected result. 
Here is an example. Struct TCB with member _state_ and the following .natvis 
```
<Type Name="TCB">
      <DisplayString Condition="state==0">Empty</DisplayString>
      <DisplayString Condition="state==128">Running</DisplayString>
      <DisplayString >{state}</DisplayString>
</Type>
```

The last _DisplayString_ is always used independently of state member value.

**Comment by pieandcakes** [comment]
@sov1178 I put an update in the `launch.json` help text and I will probably update the `launch.md` file also. I'll need to investigate the Condition statement.

**Comment by pieandcakes** [comment]
@sov1178 so I tried it with the std::vector with the following block and if I change the size of my vector from 1 to 200 to 500 I see the different DisplayStrings:
```
<Type Name="std::vector&lt;*&gt;">
      <DisplayString Condition="size==200">b</DisplayString>
      <DisplayString Condition="size==500">a</DisplayString>
      <DisplayString>{{size={_M_impl._M_finish - _M_impl._M_start}}}</DisplayString>
      <Expand>
	  <Item Name="[size]" ExcludeView="simple">_M_impl._M_finish - _M_impl._M_start</Item>
          <Item Name="[capacity]" ExcludeView="simple">_M_impl._M_end_of_storage - _M_impl._M_start</Item>
          <ArrayItems>
              <Size>_M_impl._M_finish - _M_impl._M_start</Size>
              <ValuePointer>_M_impl._M_start</ValuePointer>
          </ArrayItems>
      </Expand>
  </Type>
```
Can you enable 
```
"logging": { "trace": true, "traceResponse":true }
```
and see if you see the call to gdb to evaluate state==0 or state==128?

In my instance: 
```
1: (2104) <-1056-var-create - * "size==200"
1: (2105) ->1056^done,name="var20",numchild="0",value="false",type="bool",thread-id="1",has_more="0"
1: (2106) ->(gdb)
1: (2106) 1056: elapsed time 1
1: (2106) <-1057-var-create - * "size==500"
1: (2108) ->1057^done,name="var21",numchild="0",value="true",type="bool",thread-id="1",has_more="0"
1: (2108) ->(gdb)
```

**Comment by pieandcakes** [comment]
I wrote up an example to try and be similar to yours:
//file.natvis
```
<?xml version="1.0" encoding="utf-8"?>

<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">

    <Type Name="TCB">
        <DisplayString Condition="state==0">a</DisplayString>
        <DisplayString Condition="state==128">b</DisplayString>
        <DisplayString>{{state={state}}}</DisplayString>
    </Type>
</AutoVisualizer>
```
//main.cpp
```
#include <iostream>
#include <string.h>

using namespace std;

struct TCB {
    int state;
    string name;
};

void output(TCB* tcb)
{
    cout << "state: " <<  tcb->state << " name:" << tcb->name << endl;
}

int main (int argv, char** argc)
{
    TCB tcb;
    tcb.state = 0;
    tcb.name = "blah";
    output(&tcb);

    tcb.name = "test";
    tcb.state = 128;
    output(&tcb);

    tcb.state = 100;

    return 0;
}
```

I set a breakpoint in `output` method and the `return 0;` lineand I can see that the first call through its `a`, second call through its `b` and third time it says `{state=100}`.

**Comment by UR3IQO** [comment]
@pieandcakes thank you for the response! I set ``"engineLogging": true`` and it looks like gdb gets correct input and replies with the correct answer:
```
1: (1854) <-1026-var-create - * "((Tasks[1]).state)==0"
1: (1865) ->1026^done,name="var6",numchild="0",value="0",type="int",has_more="0"
1: (1865) ->(gdb)
1: (1865) 1026: elapsed time 11
1: (1865) <-1027-var-create - * "((Tasks[1]).state)==128"
1: (1876) ->1027^done,name="var7",numchild="0",value="1",type="int",has_more="0"
1: (1876) ->(gdb)
1: (1876) 1027: elapsed time 11
1: (1876) <-1028-var-create - * "((Tasks[1]).state)"
1: (1888) ->1028^done,name="var8",numchild="0",value="128 '\200'",type="uint8_t",has_more="0"
```
but I see the difference in return values. In my case gdb returns **0** and **1** instead of **true** and **false** (as in your case), also the type is **int** instead of **bool**. It seems it might be the source of the problem. Also I forgot to mention it is plain C not C++ code, maybe it affects how gdb handles expressions.

I got the following result in the watch window (as I understand I should get 'Running' instead of '128'):
![image](https://cloud.githubusercontent.com/assets/25171239/22769511/332d330a-ee9a-11e6-949d-48cdb88ba22f.png)

**Comment by pieandcakes** [comment]
@sov1178 So I've found where we have the check for the condition and there is a bug in the code in that we're looking for specifically a value of `"true"` to be returned from gdb from the evaluation.

[Pull Request](https://github.com/Microsoft/MIEngine/compare/master...pieandcakes:users/piel/natvisFix)

**Comment by UR3IQO** [comment]
@pieandcakes Great!

**Comment by pieandcakes** [comment]
@sov1178 this fix is in version `0.10.2` Released today
